npm/vite-plugin-cypress-esm/src/index.ts

```diff
@@ -0,0 +1,123 @@
+
+export const CypressEsm = (): Plugin => {
+  /**
+   * Remap static import calls to use the Cypress module cache
+   *
+   * ```
+   * import DefaultExport, { NamedExport, Other as Alias } from 'module'
+   * ```
+   *
+   * becomes
+   *
+   * ```
+   * import * as _cypress_module_1 from 'module';
+   * const DefaultExport = __cypressModule(_cypress_module_1);
+   * const { NamedExport, Other: Alias } = __cypressModule(_cypress_module_1);
+   * ```
+   */
+  const mapImportsToCache = (moduleId: string, code: string) => {
+    debug(`Remapping imports for module ${moduleId}`)
+    let counter = 0
+
+    const moduleIdentifier = moduleId.replace(/[^a-zA-Z\d]/g, '_')
+
+    const importRegex = /import (.+?) from (.*)/g
+
+    return code.replace(
+      importRegex,
+      (match, importVars: string, importTarget: string) => {
+        debug(`Mapping import ${counter + 1} in module ${moduleId}`)
+
+        let replacement = `import * as cypress_${moduleIdentifier}_${++counter} from ${importTarget};`
+
+        // Support `import TheDefault, { Named } from 'module'` syntax, split into two separate assignments
+        importVars.split(/(?:(?<=})\s*,\s*)|(?:\s*,\s*(?={))/g).forEach((importVar) => {
+          let destructuredImports = `const ${importVar} = ${MODULE_IMPORTER_IDENTIFIER}('${moduleId}', cypress_${moduleIdentifier}_${counter}, ${debug.enabled});`
+
+          // support `import { foo as bar } from 'module'` sytnax, converting to `const { foo: bar } ...`
+          destructuredImports = destructuredImports.replace(/(?<!\*) as /g, ': ')
+          replacement += destructuredImports
+        })
+
+        return match.replace(
+          importRegex,
+          replacement,
+        ).replace(/[\n\r]/g, '').replace(/const \* as/g, 'const')
+      },
+    )
+  }
+
+  /**
+   * Transform dynamic imports to use a runtime that turns the ES modules into proxies by injecting
+   * a `cypressModule` runtime.
+   *
+   * `cypressModule` returns a Promise than resolves to the original import that has been
+   * "proxified" to side step the fact ES Modules are immutable and sealed.
+   *
+   * Examples - https://regex101.com/r/Ic1OHA/1
+   *
+   * Given:
+   *   const m = import("./mod_1")
+   *   const m = await import("lodash")
+   *   import("./mod_2").then(mod => mod)
+   *
+   * Returns:
+   *   const m = cypressModule(import("./mod_1"))
+   *   const m = await cypressModule(import("lodash"))
+   *   cypressModule(import("./mod_2")).then(mod => mod)
+   */
+  const mapDynamicImportsToCache = (id: string, code: string) => {
+    const RE = /(import\(.+?\))/g
+
+    const c = code.replace(RE, `${MODULE_DYNAMIC_IMPORTER_IDENTIFIER}('${id}', $1, ${debug.enabled})`)
+
+    return c
+  }
+
+  return {
+    name: 'cypress:mocks',
+    enforce: 'post',
+    transform (code, id, options) {
+      // Process all files to remap imports
+      // TODO: Restrict to JS files only? Any potential for .mjs etc?
```

Comment:
Any reason not to lock this down to `.*js` files? Minor perf bump if we bypass processing json, images, etc

Reply:
Yeah, let's deal with this at some point in the future. For now I think as long as we work well with React that's probably fine!